packages/upload-api/src/upload/add.js

```diff
@@ -2,26 +2,48 @@ import pRetry from 'p-retry'
 import * as Server from '@ucanto/server'
 import * as Upload from '@web3-storage/capabilities/upload'
 import * as API from '../types.js'
+import { has as hasProvider } from '../consumer/has.js'
 
 /**
  * @param {API.UploadServiceContext} context
  * @returns {API.ServiceMethod<API.UploadAdd, API.UploadAddOk, API.Failure>}
  */
-export function uploadAddProvider({ access, uploadTable, dudewhereBucket }) {
+export function uploadAddProvider(context) {
   return Server.provide(Upload.add, async ({ capability, invocation }) => {
+    const { uploadTable, dudewhereBucket, signer } = context
+    const serviceDID = /** @type {import('../types.js').ProviderDID} */ (
+      signer.did()
+    )
     const { root, shards } = capability.nb
-    const space = Server.DID.parse(capability.with)
+    const space = /** @type {import('@ucanto/interface').DIDKey} */ (
+      Server.DID.parse(capability.with).did()
+    )
     const issuer = invocation.issuer.did()
-    const allocated = await access.allocateSpace(invocation)
-
-    if (allocated.error) {
-      return allocated
+    const hasProviderResult = await hasProvider(
```

Comment:
Do we not want to check the block list here ?

I also would recommend command style API over defensive style API that is instead of having `hasProvider` you could have `useCapability` so that you don't have to have one branch for propagating an error and other branch for creating an error if answer to question was `false`.

That is also why we had `allocateSpace` as opposed to `hasSpace.

P.S. It is very same idea as with [Parse, don’t validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) _(a good read if you have not come across it before)_.

Reply:
oo I like this, and I also like being able to re-use `allocate` the way we were - I've pushed a new commit that makes this happen, and drops the `size` input on `space/allocate` - we weren't using it for anything and the fact that we don't have a `size` in the `upload/add` provider was the main reason I moved away from it.

I suspect we may want to re-introduce it later, either as an optional parameter that triggers a check against a space's allocation or as a separate but similar capability, but this seems ok to me for now - let me know if you think it might be a problem!

https://github.com/web3-storage/w3up/pull/832/commits/22b7a69e9861960206456b9b9802727b44cc79ec